glib2/ext/glib2/rbglib-variant.c

```diff
@@ -39,11 +39,39 @@ rbg_variant_to_ruby(GVariant *variant)
 
     if (g_variant_type_equal(type, G_VARIANT_TYPE_BOOLEAN)) {
         return CBOOL2RVAL(g_variant_get_boolean(variant));
-    } else if (g_variant_type_equal(type, G_VARIANT_TYPE_STRING)) {
+    } else if (g_variant_type_equal(type, G_VARIANT_TYPE_BYTE)) {
+        return UINT2NUM(g_variant_get_byte(variant));
+    } else if (g_variant_type_equal(type, G_VARIANT_TYPE_INT16)) {
+        return INT2NUM(g_variant_get_int16(variant));
+    } else if (g_variant_type_equal(type, G_VARIANT_TYPE_UINT16)) {
+        return UINT2NUM(g_variant_get_uint16(variant));
+    } else if (g_variant_type_equal(type, G_VARIANT_TYPE_INT32)) {
+        return INT2NUM(g_variant_get_int32(variant));
+    } else if (g_variant_type_equal(type, G_VARIANT_TYPE_UINT32)) {
+        return UINT2NUM(g_variant_get_uint32(variant));
+    } else if (g_variant_type_equal(type, G_VARIANT_TYPE_INT64)) {
+        return rbglib_int64_to_num(g_variant_get_int64(variant));
+    } else if (g_variant_type_equal(type, G_VARIANT_TYPE_UINT64)) {
+        return rbglib_uint64_to_num(g_variant_get_uint64(variant));
+    } else if (g_variant_type_equal(type, G_VARIANT_TYPE_DOUBLE)) {
+        return rb_float_new(g_variant_get_double(variant));
+    } else if (g_variant_type_equal(type, G_VARIANT_TYPE_STRING) ||
+               g_variant_type_equal(type, G_VARIANT_TYPE_OBJECT_PATH) ||
+               g_variant_type_equal(type, G_VARIANT_TYPE_SIGNATURE)) {
         const gchar *string;
         gsize string_length;
         string = g_variant_get_string(variant, &string_length);
         return CSTR2RVAL_LEN(string, string_length);
+    } else if (g_variant_type_equal(type, G_VARIANT_TYPE_HANDLE)) {
+        return INT2NUM(g_variant_get_handle(variant));
+    } else if (g_variant_type_equal(type, G_VARIANT_TYPE_VARIANT)) {
+        return rbg_variant_new(g_variant_get_variant(variant));
+    } else if (g_variant_type_is_array(type)) {
+        gsize len = g_variant_n_children(variant);
+        VALUE ary = rb_ary_new2(len);
+        for (gsize i = 0; i < len; i++)
```

Comment:
On Thu, Jan 14, 2016 at 02:49:18AM -0800, Kouhei Sutou wrote:
> >          const gchar *string;
> >          gsize string_length;
> >          string = g_variant_get_string(variant, &string_length);
> >          return CSTR2RVAL_LEN(string, string_length);
> > +    } else if (g_variant_type_equal(type, G_VARIANT_TYPE_HANDLE)) {
> > +        return INT2NUM(g_variant_get_handle(variant));
> > +    } else if (g_variant_type_equal(type, G_VARIANT_TYPE_VARIANT)) {
> > +        return rbg_variant_new(g_variant_get_variant(variant));
> > +    } else if (g_variant_type_is_array(type)) {
> > +        gsize len = g_variant_n_children(variant);
> > +        VALUE ary = rb_ary_new2(len);
> > +        for (gsize i = 0; i < len; i++)
> 
> Don't define variable in `for`.

OK, fixed.
But it is a bit sad to explicitely avoid some nice language features
that are part of the ISO C standard since more than 17 years now ! (C99)

Moreover, now that gcc5 defaults to C11 (while gcc <= 4.9 defaults to
C90), we should probably expicitely specify in the build system what
version of the C standard we are using. Without this, the code is
currently compiled following different standard depending on the compiler
used.
I would suggest to add a -std=c11 to the compiler options. (This is
supported since gcc 4.7 which is already 4 years old)